Python/Product/Analysis/Interpreter/Ast/AstAnalysisFunctionWalker.cs

```diff
@@ -111,25 +102,139 @@ FunctionDefinition targetFunction
+                    _scope.SetInScope(ne.Name, value);
+                }
+
+                // Tuple = Tuple. Transfer values.
+                if (lhs is TupleExpression tex) {
+                    if (value is TupleExpression valTex) {
+                        var returnedExpressions = valTex.Items.ToArray();
+                        var names = tex.Items.Select(x => (x as NameExpression)?.Name).ToArray();
+                        for (var i = 0; i < Math.Min(names.Length, returnedExpressions.Length); i++) {
+                            if (returnedExpressions[i] != null) {
+                                var v = _scope.GetValueFromExpression(returnedExpressions[i]);
+                                _scope.SetInScope(names[i], v);
+                            }
+                        }
+                        continue;
+                    }
+
+                    // Tuple = 'tuple value' (such as from callable). Transfer values.
+                    if (value is AstPythonConstant c && c.Type is AstPythonTuple tuple) {
+                        var types = tuple.Types.ToArray();
+                        var names = tex.Items.Select(x => (x as NameExpression)?.Name).ToArray();
+                        for (var i = 0; i < Math.Min(names.Length, types.Length); i++) {
+                            if (names[i] != null) {
+                                _scope.SetInScope(names[i], new AstPythonConstant(types[i]));
+                            }
+                        }
+                    }
+                }
+            }
+            return base.Walk(node);
+        }
+
+        public override bool Walk(IfStatement node) {
+            // Handle basic check such as
+            // if isinstance(value, type):
+            //    return value
+            // by assigning type to the value unless clause is raising exception.
+            var ce = node.Tests.FirstOrDefault()?.Test as CallExpression;
+            if (ce?.Target is NameExpression ne && ne?.Name == "isinstance" && ce.Args.Count == 2) {
+                var name = (ce.Args[0].Expression as NameExpression)?.Name;
+                var typeName = (ce.Args[1].Expression as NameExpression)?.Name;
+                if (name != null && typeName != null) {
+                    var typeId = BuiltinTypeIdFromName(typeName);
+                    if (typeId != BuiltinTypeId.Unknown) {
+                        _scope.SetInScope(name, new AstPythonConstant(new AstPythonBuiltinType(typeName, typeId)));
+                    }
+                }
+            }
             return base.Walk(node);
         }
 
         public override bool Walk(ReturnStatement node) {
-            foreach (var type in _scope.GetTypesFromValue(_scope.GetValueFromExpression(node.Expression))) {
-                _returnTypes.Add(type);
+            var typesAdded = false;
+            if (node.Expression is NameExpression nex && nex.Name == "self") {
```

Comment:
Is this actually necessary? Haven't we already set `self` to the correct value earlier?

Reply:
Hmm... looks like we haven't, which means we won't properly interpret `x = self.something; return x` now?

Reply:
This is supposed to apply only to `return self` so we return type without private members. In fact, `typesAdded` is not necessary, we can just return.